File: reference/goathal/pqsign.py
"""
Post-quantum signature interface (PROTOCOL layer — device-agnostic).

The protocol is written against the algorithm-agnostic `Signer` / `verify` interface.
Production uses ML-DSA-65 (FIPS 204) via the proper Rust crate; the wire format
length-prefixes signatures and public keys so the exact algorithm and its byte sizes
are irrelevant to serialization.

Reference note: no ML-DSA library is available in this environment, so the reference
`ReferenceSigner` is backed by Ed25519 (via `cryptography`) purely to exercise GENUINE
asymmetric sign/verify semantics (round-trip + tamper detection) in tests. It is a
stand-in ONLY — it is NOT post-quantum. `AlgId.ML_DSA_65` and the size constants below
model the production algorithm so bandwidth math stays faithful.
"""
from abc import ABC, abstractmethod
from enum import Enum

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey, Ed25519PublicKey,
)
from cryptography.hazmat.primitives import serialization
from cryptography.exceptions import InvalidSignature


class AlgId(Enum):
    REFERENCE_ED25519 = 0     # reference stand-in only (NOT post-quantum)
    ML_DSA_65 = 1             # production target (FIPS 204)


# Production ML-DSA-65 sizes (bytes) — for bandwidth documentation; the wire format is
# length-prefixed, so real signatures (much larger than the reference) drop in unchanged.
ML_DSA_65_PUBKEY_BYTES = 1952
ML_DSA_65_SIG_BYTES = 3309


class Signer(ABC):
    @property
    @abstractmethod
    def alg_id(self) -> AlgId: ...

    @abstractmethod
    def public_key(self) -> bytes: ...

    @abstractmethod
    def sign(self, msg: bytes) -> bytes: ...


def verify(alg_id: AlgId, pubkey: bytes, msg: bytes, sig: bytes) -> bool:
    """Algorithm-dispatched verification using ONLY the public key."""
    if alg_id == AlgId.REFERENCE_ED25519:
        try:
            Ed25519PublicKey.from_public_bytes(pubkey).verify(sig, msg)
            return True
        except (InvalidSignature, ValueError):
            return False
    if alg_id == AlgId.ML_DSA_65:
        raise NotImplementedError(
            "ML-DSA-65 verify requires the production PQC crate; not available in the "
            "reference environment. Swap in the real backend behind this same function."
        )
    raise ValueError(f"unknown alg_id {alg_id}")


class ReferenceSigner(Signer):
    """Ed25519-backed stand-in for ML-DSA-65 (NOT post-quantum). Reference/testing only."""

    def __init__(self, sk: Ed25519PrivateKey = None):
        self._sk = sk or Ed25519PrivateKey.generate()

    @property
    def alg_id(self) -> AlgId:
        return AlgId.REFERENCE_ED25519

    def public_key(self) -> bytes:
        return self._sk.public_key().public_bytes(
            serialization.Encoding.Raw, serialization.PublicFormat.Raw)

    def sign(self, msg: bytes) -> bytes:
        return self._sk.sign(msg)


def new_reference_keypair() -> ReferenceSigner:
    return ReferenceSigner()

File: goatcoin-rs/crates/goat-protocol/src/pqsign.rs
//! Post-quantum signatures (device-agnostic). WP-0.2.
//!
//! Production ML-DSA-65 (FIPS 204) via the RustCrypto `ml-dsa` crate, behind an
//! algorithm-agnostic `Signer` / `verify` interface. The Ed25519 reference stand-in used
//! during the Python reference (R-CAP1) is GONE — this is a genuine post-quantum signer.
//! The wire format length-prefixes signatures and public keys, so the algorithm's byte
//! sizes (pubkey 1952 B, sig 3309 B) are irrelevant to serialization.

use ml_dsa::signature::{Signer as _, Verifier as _};
use ml_dsa::{
    EncodedSignature, EncodedVerifyingKey, Generate, Keypair as _, MlDsa65, Signature, SigningKey,
    VerifyingKey,
};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum AlgId {
    MlDsa65,
}

impl AlgId {
    pub fn as_u16(self) -> u16 {
        match self {
            AlgId::MlDsa65 => 1,
        }
    }
}

/// ML-DSA-65 sizes (FIPS 204). Asserted by tests against the live crate.
pub const ML_DSA_65_PUBKEY_BYTES: usize = 1952;
pub const ML_DSA_65_SIG_BYTES: usize = 3309;

/// Algorithm-agnostic signing surface used by the protocol.
pub trait Signer {
    fn alg_id(&self) -> AlgId;
    fn public_key(&self) -> Vec<u8>;
    fn sign(&self, msg: &[u8]) -> Vec<u8>;
}

/// Verify with ONLY the public key, dispatched by algorithm.
pub fn verify(alg: AlgId, pubkey: &[u8], msg: &[u8], sig: &[u8]) -> bool {
    match alg {
        AlgId::MlDsa65 => {
            let enc_vk = match EncodedVerifyingKey::<MlDsa65>::try_from(pubkey) {
                Ok(e) => e,
                Err(_) => return false,
            };
            let vk = VerifyingKey::<MlDsa65>::decode(&enc_vk);
            let enc_sig = match EncodedSignature::<MlDsa65>::try_from(sig) {
                Ok(e) => e,
                Err(_) => return false,
            };
            let signature = match Signature::<MlDsa65>::decode(&enc_sig) {
                Some(s) => s,
                None => return false,
            };
            vk.verify(msg, &signature).is_ok()
        }
    }
}

/// Real ML-DSA-65 signer.
pub struct MlDsaSigner {
    sk: SigningKey<MlDsa65>,
}

impl MlDsaSigner {
    pub fn generate() -> Self {
        Self {
            sk: SigningKey::<MlDsa65>::generate(),
        }
    }
}

impl Signer for MlDsaSigner {
    fn alg_id(&self) -> AlgId {
        AlgId::MlDsa65
    }
    fn public_key(&self) -> Vec<u8> {
        let vk: VerifyingKey<MlDsa65> = self.sk.verifying_key();
        vk.encode().as_slice().to_vec()
    }
    fn sign(&self, msg: &[u8]) -> Vec<u8> {
        let sig: Signature<MlDsa65> = self.sk.sign(msg);
        sig.encode().as_slice().to_vec()
    }
}
